pkg/utils/db.go

```diff
@@ -277,35 +277,140 @@ func IsErrBinlogPurged(err error) bool {
+	return nil
+}
+
+// NormalRetryOperation will retry 100 times
+func (conn *BaseConn) NormalRetryOperation(ctx *tcontext.Context, operateFn func(*tcontext.Context, int) (interface{}, error), retryFn func(int, error) bool) (interface{}, error) {
+	var err error
+	var ret interface{}
+	for i := 0; i < 100; i++ {
+		ret, err = operateFn(ctx, i)
 		if err != nil {
-			return nil, errors.Annotatef(err, "can not decrypt password %s of user %s for db %s:%d", password, user, host, port)
+			if retryFn(i, err) {
+				continue
+			}
 		}
+		break
 	}
+	return ret, err
+}
+
+// InfinityRetryOperation retry operation until succeed
+func (conn *BaseConn) InfinityRetryOperation(operateFn func() (interface{}, error), retryFn func(error) bool) (interface{}, error) {
+	var err error
+	var ret interface{}
+	for {
+		ret, err = operateFn()
+		if err != nil {
+			if retryFn(err) {
+				continue
+			}
+		}
+		break
+	}
+	return ret, err
+}
+
+// QuerySQL query sql
+func (conn *BaseConn) QuerySQL(tctx *tcontext.Context, query string) (*sql.Rows, error) {
+	if conn == nil || conn.DB == nil {
+		return nil, errors.NotValidf("database connection")
+	}
+
+	var (
+		err  error
+		rows *sql.Rows
+	)
+
+	tctx.L().Debug("Query statement", zap.String("Query", query))
+
+	rows, err = conn.DB.QueryContext(tctx.Context(), query)
 
-	dbDSN := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8&interpolateParams=true&readTimeout=%s", user, password, host, port, timeout)
-	db, err = sql.Open("mysql", dbDSN)
 	if err != nil {
+		tctx.L().Error("Query statement failed", zap.String("Query", query), log.ShortError(err))
 		return nil, errors.Trace(err)
 	}
+	return rows, nil
+}
+
+// ExecuteSQL execute sql
+func (conn *BaseConn) ExecuteSQL(tctx *tcontext.Context, sqls []SQL) (int, error) {
+	if len(sqls) == 0 {
+		return 0, nil
+	}
+
+	if conn == nil || conn.DB == nil {
+		return 0, errors.NotValidf("database connection")
+	}
+
+	if affected, err := conn.executeSQLImp(tctx, sqls); err != nil {
+		return affected, err
+	}
+	return 0, nil
+}
+
+func (conn *BaseConn) executeSQLImp(tctx *tcontext.Context, sqls []SQL) (int, error) {
+	if conn == nil || conn.DB == nil {
+		return 0, errors.NotValidf("database connection")
+	}
 
-	return
-}*/
+	txn, err := conn.DB.Begin()
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+
+	l := len(sqls)
+
+	for i := range sqls {
+		tctx.L().Debug("execute statement", zap.String("Query", sqls[i].Query), zap.Reflect("argument", sqls[i].Args))
+
+		_, err = txn.ExecContext(tctx.Context(), sqls[i].Query, sqls[i].Args...)
+		if err != nil {
+			tctx.L().Error("execute statement failed", zap.String("Query", sqls[i].Query), zap.Reflect("argument", sqls[i].Args), log.ShortError(err))
+			rerr := txn.Rollback()
+			if rerr != nil {
+				tctx.L().Error("rollback failed", zap.String("Query", sqls[i].Query), zap.Reflect("argument", sqls[i].Args), log.ShortError(rerr))
+			}
+			// we should return the exec err, instead of the rollback rerr.
+			return i, errors.Trace(err)
```

Comment:
should we use the new error system in #216

Reply:
this PR need rebase master after #216 merged